### PR TITLE
profile & connections routes & logic 

### PR DESCRIPTION
**Pull request by davideastmond**
This covers routes and associated logic

- getting a profile (GET `/users/:id`). if id is `me` then this route retrieves requestor's own profile.
- getting connections (GET `users/:id/connections`) me = retrieving requestor's own connections

- adding a new connection (PUT `connections/:id`)

Resolves #7 #12 and #33

### CODE REVIEW

File: api/src/routes/users.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+const router = express.Router();
+
+router.get("/:id", routeProtector, [ param("id").not().isEmpty().trim().escape()], async (req: any, res: Response) => {
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
+    return res.status(400).send({ errors: errors.array() });
+  }
+  if (req.params.id === "me") {
+    const homeProfileData = await getProfileById(req.user._id);
+    return res.status(200).send({ data: homeProfileData });
```
Comment (lily-law) [inline api/src/routes/users.ts:17]:
```suggestion
    return res.status(200).send(homeProfileData);
```

File: api/src/routes/users.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  } else {
+    const otherUserData = await getProfileById(req.params.id);
+    return res.status(200).send({ data: otherUserData });
```
Comment (lily-law) [inline api/src/routes/users.ts:20]:
```suggestion
    return res.status(200).send(otherUserData);
```

File: api/src/routes/users.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+routeProtector,
+[ param("id").not().isEmpty().trim().escape()],
+async (req: any, res: Response) => {
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
+    return res.status(500).send({ errors: errors.array() });
+  }
+  if (req.params.id === "me") {
+    // Get requesting user's connections
+    res.status(200).send({ data: req.user.connections});
```
Comment (lily-law) [inline api/src/routes/users.ts:34]:
```suggestion
    res.status(200).send(req.user.connections);
```

File: api/src/routes/users.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  } else {
+    // get and respond with connections of some other user
+    try {
+      const otherUser = await UserModel.findById(req.params.id);
+      if (otherUser) {
+        res.status(200).send({ data: otherUser.connections });
```
Comment (lily-law) [inline api/src/routes/users.ts:40]:
```suggestion
        res.status(200).send(otherUser.connections);
```

**Review by lily-law — APPROVED** [review]
Looking good, just the suggested changes. When we receive the response on the frontend what's in send is already wrapped in data, so currently we get res: {data: {data: {...userStuff}}, ...otherResStuff}

File: api/src/middleware/login-validator.ts [hunk 5]
```diff
@@ -1,11 +1,14 @@
...
+import { createError } from "../utils/errors";
 
 // Checks if user already has a session
 export function checkNotAuthenticated (req: Request, res: Response, next: NextFunction) {
   if (!req.user) {
     next();
   } else {
     // Redirect to a profile page as user is already logged in
-    res.status(400).send({ message: "There is already a session. You need to log out."}); // Placeholder
+    res.status(400).send({ errors: [ {
```
Comment (lily-law) [inline api/src/middleware/login-validator.ts:10]: Is there a reason to give an error here? Like could there be a situation where the client would need to be aware that they already have a session or would it be better to just bypass and respond 200/OK so as far as the client knows they've authed successfully?

File: api/src/models/user/user.methods.ts [hunk 6]
```diff
@@ -93,8 +93,10 @@ export async function findOneByEncryptedEmail (this: IUserModel, encryptedEmail:
  * @param this *
  * @param objId object id
  */
-export async function addConnectionToUser (this: IUserDocument, objId: string, isTeamMate?: boolean): Promise<[IUserDocument, IUserDocument]> {
+export async function addConnectionToUser (this: IUserDocument, objId: string, isTeamMate?: boolean): Promise<IUserDocument> {
```
Comment (lily-law) [inline api/src/models/user/user.methods.ts:106]:
```suggestion
export async function addConnectionToUser (thisUsersDocument: IUserDocument, objId: string, isTeamMate?: boolean): Promise<IUserDocument> {
```

Comment (lily-law) [reply api/src/models/user/user.methods.ts:106]: Then we won't need a comment to explain?

File: api/src/models/user/user.methods.ts [hunk 7]
```diff
@@ -105,17 +107,42 @@ export async function addConnectionToUser (this: IUserDocument, objId: string, i
       targetUser["connectionOf"][this._id] = originatorConnection;
 
       // Saves the changes
-      const originator = await UserModel.findByIdAndUpdate(this.id, this);
-      await UserModel.findByIdAndUpdate(objId, targetUser);
+      await this.save();
```
Comment (lily-law) [inline api/src/models/user/user.methods.ts:118]:
```suggestion
      await thisUsersDocument.save();
```

Comment (lily-law) [reply api/src/models/user/user.methods.ts:118]: What you've done with `this` looks much cleaner then originator... let's remove the commented out originator stuff, unless I'm missing something?

File: api/src/utils/errors.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+export type AppError = {
```
Comment (lily-law) [inline api/src/utils/errors.ts:1]: Is this needed? maybe the standard Error constructor take care of this and more https://nodejs.org/api/errors.html#errors_new_error_message

**Review by lily-law — COMMENTED** [review]
Nice, connections are looking good, I need to catch up on that for the frontend... a few changes checkout but otherwise really good 👍

File: api/src/utils/errors.ts (hunk 8 above)
Comment (davideastmond) [reply api/src/utils/errors.ts:1]: We'll come back to this refactor if needed

**Review by lily-law — APPROVED** [review]

**Pull request merged**